### PR TITLE
[backport 17.06] NetworkDB node management race

### PR DESCRIPTION
**Pull request by fcrisciani**
The previous logic was not properly handling the case of a node
that was failing and coming back in short period of time.
The issue was in the handling of the network messages.
When a node joins it sync with other nodes, these are passing
the whole list of nodes that at best of their knowledge are part
of a network. At this point if the node receives that node A is part
of the network it saves it before having received the notification
that node A is actually alive (coming from memberlist).
If node A failed the source node will receive the notification
while the new joined node won't because memberlist never advertise
node A as available. In this case the new node will never purge
node A from its state but also worse, will accept any table notification
where node A is the owner and so will end up in a out of sync state
with the rest of the cluster.

Signed-off-by: Flavio Crisciani <flavio.crisciani@docker.com>
(cherry picked from commit 4702ccbbd20cc3827ba112664e3f502f00d32660)

### CODE REVIEW

**Comment by fcrisciani** [comment]
Would like to give it a test before merging it considering that is pretty different from the master patch

**Comment by thaJeztah** [comment]
Can you update the commit message with the description you put in the PR?

File: networkdb/delegate.go [hunk 1]
```diff
@@ -28,12 +28,13 @@ func (nDB *NetworkDB) getNode(nEvent *NodeEvent) *node {
 	} {
 		if n, ok := nodes[nEvent.NodeName]; ok {
 			if n.ltime >= nEvent.LTime {
-				return nil
+				return false, nil
 			}
-			return n
+			found := &nodes == &nDB.nodes
```
Comment (fcrisciani) [inline networkdb/delegate.go]: this need to be changed

**Comment by fcrisciani** [comment]
@mavenugo @ddebroy can you guys take a look?

**Comment by ddebroy** [comment]
LGTM

**Comment by mavenugo** [comment]
@fcrisciani Its hard to justify that this PR is a backport of https://github.com/docker/libnetwork/pull/2013 . Also it isnt clear on how and why other changes in https://github.com/docker/libnetwork/pull/2013 such as `PurgeReincarnation` isnt required for the backports. Can you pls give some explanations ?

**Comment by fcrisciani** [comment]
@mavenugo the 2 codebase diverged long time ago. One of the main difference is the memberlist node name. While on master the node name length is a fixed 8 character random name, here it is still a composition with the hostname.
This patch being a backport of a specific fix tries to address it with minimal change avoiding any code optimization that will simply increase risk of regression.

**Comment by mavenugo** [comment]
@fcrisciani okay. got it.

LGTM

**Pull request merged**